# Gesundheitsausgaben-Analyse

## Projektübersicht
Dieses Projekt analysiert Gesundheitsausgaben basierend auf verschiedenen Kategorien wie Finanzierungsquellen, Arten von Gesundheitsleistungen und Zeiträumen.

Ziel ist es, Einblicke in die Entwicklung der Gesundheitskosten zu gewinnen und Muster zu identifizieren, die zukünftige Entscheidungen in der Gesundheitspolitik beeinflussen können.

### **Projektziele:**
1. **Datenbereinigung:** Aufbereitung der Rohdaten durch Entfernen unnötiger Spalten.
2. **Datenanalyse:** Identifikation von Trends in den Gesundheitsausgaben.
3. **Visualisierung:** Erstellung von Diagrammen für die Interpretation.
4. **Berichterstellung:** Zusammenfassung der wichtigsten Erkenntnisse.

### **Verwendete Datensätze:**
Die Analyse basiert auf folgenden CSV-Dateien:
- `OGD_gesausgaben01_HVD_HCHF_1_C-HCGES-0.csv` (Gesundheitsleistungen und -güter)
- `OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0.csv` (Zeitreihen der Gesundheitsausgaben)
- `OGD_gesausgaben01_HVD_HCHF_1_HEADER.csv` (Metadaten über Spaltenbezeichnungen)

### **Projektstruktur:**
- `data/` → Enthält die Anfangsdaten
- `notebooks/` → Enthält das Jupyter Notebook
- `output/` → Speichert bereinigte und aggregierte Daten
- `scripts/` → Enthält wiederverwendbare Python-Funktionen

## Libraries

In [163]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

## 1. Datenanalyse

1. **Hauptdatensatz:**  
   - Datei: `OGD_gesausgaben01_HVD_HCHF_1.csv`  
   - Enthält die eigentlichen Gesundheitsausgaben-Daten.  

2. **Hilfsdatensätze:**  
   - `OGD_gesausgaben01_HVD_HCHF_1_C-HCGES-0.csv` – Gesundheitsleistungen und -güter (HC)  
   - `OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0.csv` – Zeitreihen der Gesundheitsausgaben  
   - `OGD_gesausgaben01_HVD_HCHF_1_HEADER.csv` – Header-Metadaten zur Beschreibung der Spalten  

#### Warum verwenden wir `;` als Delimiter?  

Die bereitgestellten CSV-Dateien verwenden das Semikolon (`;`) als Trennzeichen (wir glauben um Verwechslungen mit Dezimalpunkten oder -kommas zu vermeiden)

Daher muss beim Einlesen `delimiter=';'` explizit angegeben werden, um die Daten korrekt zu parsen.  


In [164]:
def read_dfs(file_path="data"):
    dataframes = {}
    for filename in os.listdir(file_path):
        if not filename.endswith('.csv'):
            continue
        file_path = os.path.join('data', filename)
        df_ = pd.read_csv(file_path, delimiter=';')
        df_name = filename.replace('.csv', '')
        dataframes[df_name] = df_
    return dataframes

dataframes = read_dfs()
df_main = dataframes['OGD_gesausgaben01_HVD_HCHF_1']
df_hc = dataframes['OGD_gesausgaben01_HVD_HCHF_1_C-HCGES-0']
df_time = dataframes['OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0']
df_header = dataframes['OGD_gesausgaben01_HVD_HCHF_1_HEADER']

def analyze_dataframe(df, name):
    print(f"\n--- {name} ---")
    
    print("\nErste 5 Zeilen:")
    display(df.head())

    # zufällige Stichprobe
    print("\nZufällige Stichprobe:")
    display(df.sample(5))

    # Überblick über die Struktur des DataFrames
    print("\nDateninfo:")
    df.info()

    # Deskriptive Statistiken
    print("\nDeskriptive Statistiken (numerische und kategoriale Werte):")
    display(df.describe())

    # Unique Werte für jede Spalte (nur für kategorische Spalten sinnvoll)
    print("\nAnzahl eindeutiger Werte pro Spalte:")
    for col in df.columns:
        unique_values = df[col].nunique()
        print(f"{col}: {unique_values} eindeutige Werte")
        # print(f"  Werte: {df[col].unique()}")  # unnötig zu viel output
        
    # ganz wichtig noch null werte!
    print(f"Fehlende Werte in {name}:")
    display(df.isnull().sum())

for df_name, df in dataframes.items():
    analyze_dataframe(df, df_name)


--- OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0 ---

Erste 5 Zeilen:


code  name  Unnamed: 2  en_name  de_desc  de_link  en_desc  \
0  ZEITGES-2004  2004         NaN     2004      NaN      NaN      NaN   
1  ZEITGES-2005  2005         NaN     2005      NaN      NaN      NaN   
2  ZEITGES-2006  2006         NaN     2006      NaN      NaN      NaN   
3  ZEITGES-2007  2007         NaN     2007      NaN      NaN      NaN   
4  ZEITGES-2008  2008         NaN     2008      NaN      NaN      NaN   

   en_link  de_syn  en_syn  
0      NaN     NaN     NaN  
1      NaN     NaN     NaN  
2      NaN     NaN     NaN  
3      NaN     NaN     NaN  
4      NaN     NaN     NaN


Zufällige Stichprobe:


code  name  Unnamed: 2  en_name  de_desc  de_link  en_desc  \
8   ZEITGES-2012  2012         NaN     2012      NaN      NaN      NaN   
14  ZEITGES-2018  2018         NaN     2018      NaN      NaN      NaN   
0   ZEITGES-2004  2004         NaN     2004      NaN      NaN      NaN   
15  ZEITGES-2019  2019         NaN     2019      NaN      NaN      NaN   
1   ZEITGES-2005  2005         NaN     2005      NaN      NaN      NaN   

    en_link  de_syn  en_syn  
8       NaN     NaN     NaN  
14      NaN     NaN     NaN  
0       NaN     NaN     NaN  
15      NaN     NaN     NaN  
1       NaN     NaN     NaN


Dateninfo:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19 entries, 0 to 18
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   code        19 non-null     category
 1   name        19 non-null     int64   
 2   Unnamed: 2  0 non-null      float64 
 3   en_name     19 non-null     int64   
 4   de_desc     0 non-null      float64 
 5   de_link     0 non-null      float64 
 6   en_desc     0 non-null      float64 
 7   en_link     0 non-null      float64 
 8   de_syn      0 non-null      float64 
 9   en_syn      0 non-null      float64 
dtypes: category(1), float64(7), int64(2)
memory usage: 2.2 KB

Deskriptive Statistiken (numerische und kategoriale Werte):


name  Unnamed: 2      en_name  de_desc  de_link  en_desc  \
count    19.000000         0.0    19.000000      0.0      0.0      0.0   
mean   2013.000000         NaN  2013.000000      NaN      NaN      NaN   
std       5.627314         NaN     5.627314      NaN      NaN      NaN   
min    2004.000000         NaN  2004.000000      NaN      NaN      NaN   
25%    2008.500000         NaN  2008.500000      NaN      NaN      NaN   
50%    2013.000000         NaN  2013.000000      NaN      NaN      NaN   
75%    2017.500000         NaN  2017.500000      NaN      NaN      NaN   
max    2022.000000         NaN  2022.000000      NaN      NaN      NaN   

       en_link  de_syn  en_syn  
count      0.0     0.0     0.0  
mean       NaN     NaN     NaN  
std        NaN     NaN     NaN  
min        NaN     NaN     NaN  
25%        NaN     NaN     NaN  
50%        NaN     NaN     NaN  
75%        NaN     NaN     NaN  
max        NaN     NaN     NaN


Anzahl eindeutiger Werte pro Spalte:
code: 19 eindeutige Werte
name: 19 eindeutige Werte
Unnamed: 2: 0 eindeutige Werte
en_name: 19 eindeutige Werte
de_desc: 0 eindeutige Werte
de_link: 0 eindeutige Werte
en_desc: 0 eindeutige Werte
en_link: 0 eindeutige Werte
de_syn: 0 eindeutige Werte
en_syn: 0 eindeutige Werte
Fehlende Werte in OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0:


code           0
name           0
Unnamed: 2    19
en_name        0
de_desc       19
de_link       19
en_desc       19
en_link       19
de_syn        19
en_syn        19
dtype: int64


--- OGD_gesausgaben01_HVD_HCHF_1_HEADER ---

Erste 5 Zeilen:


code                                               name  \
0  C-ZEITGES-0                                               Zeit   
1    C-HCGES-0                   Gesundheitsleistungen_Güter (HC)   
2        F-HF1  HF.1 Finanzierungssystem Staat und verpflichte...   
3       F-HF11                   HF.1.1 Finanzierungssystem Staat   
4       F-HF12  HF.1.2 Verpflichtende Beitragssysteme: Kranken...   

                                             en_name  Unnamed: 3  Unnamed: 4  \
0                                               Time         NaN         NaN   
1                         Health care functions (HC)         NaN         NaN   
2  HF.1 Government schemes and compulsory contrib...         NaN         NaN   
3                          HF.1.1 Government schemes         NaN         NaN   
4  HF.1.2 Compulsory contributory health insuranc...         NaN         NaN   

   Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9  
0         NaN         NaN         NaN         NaN         NaN  
1         NaN         NaN         NaN         NaN         NaN  
2         NaN         NaN         NaN         NaN         NaN  
3         NaN         NaN         NaN         NaN         NaN  
4         NaN         NaN         NaN         NaN         NaN


Zufällige Stichprobe:


code                                               name  \
6   F-HF122  HF.1.2.2 Verpflichtende private Krankenversich...   
10   F-HF23             HF.2.3 Finanzierungssystem Unternehmen   
5   F-HF121  HF.1.2.1 Finanzierungssystem Sozialversicherun...   
3    F-HF11                   HF.1.1 Finanzierungssystem Staat   
13   F-HF32   HF.3.2 Selbstbehalte (Kostenteilung mit Dritten)   

                                          en_name  Unnamed: 3  Unnamed: 4  \
6   HF.1.2.2 Compulsory private insurance schemes         NaN         NaN   
10            HF.2.3 Enterprise financing schemes         NaN         NaN   
5        HF.1.2.1 Social health insurance schemes         NaN         NaN   
3                       HF.1.1 Government schemes         NaN         NaN   
13    HF.3.2 Cost-sharing with third-party payers         NaN         NaN   

    Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9  
6          NaN         NaN         NaN         NaN         NaN  
10         NaN         NaN         NaN         NaN         NaN  
5          NaN         NaN         NaN         NaN         NaN  
3          NaN         NaN         NaN         NaN         NaN  
13         NaN         NaN         NaN         NaN         NaN


Dateninfo:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16 entries, 0 to 15
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   code        16 non-null     category
 1   name        16 non-null     category
 2   en_name     16 non-null     category
 3   Unnamed: 3  0 non-null      float64 
 4   Unnamed: 4  0 non-null      float64 
 5   Unnamed: 5  0 non-null      float64 
 6   Unnamed: 6  0 non-null      float64 
 7   Unnamed: 7  0 non-null      float64 
 8   Unnamed: 8  0 non-null      float64 
 9   Unnamed: 9  0 non-null      float64 
dtypes: category(3), float64(7)
memory usage: 3.1 KB

Deskriptive Statistiken (numerische und kategoriale Werte):


Unnamed: 3  Unnamed: 4  Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  \
count         0.0         0.0         0.0         0.0         0.0         0.0   
mean          NaN         NaN         NaN         NaN         NaN         NaN   
std           NaN         NaN         NaN         NaN         NaN         NaN   
min           NaN         NaN         NaN         NaN         NaN         NaN   
25%           NaN         NaN         NaN         NaN         NaN         NaN   
50%           NaN         NaN         NaN         NaN         NaN         NaN   
75%           NaN         NaN         NaN         NaN         NaN         NaN   
max           NaN         NaN         NaN         NaN         NaN         NaN   

       Unnamed: 9  
count         0.0  
mean          NaN  
std           NaN  
min           NaN  
25%           NaN  
50%           NaN  
75%           NaN  
max           NaN


Anzahl eindeutiger Werte pro Spalte:
code: 16 eindeutige Werte
name: 16 eindeutige Werte
en_name: 16 eindeutige Werte
Unnamed: 3: 0 eindeutige Werte
Unnamed: 4: 0 eindeutige Werte
Unnamed: 5: 0 eindeutige Werte
Unnamed: 6: 0 eindeutige Werte
Unnamed: 7: 0 eindeutige Werte
Unnamed: 8: 0 eindeutige Werte
Unnamed: 9: 0 eindeutige Werte
Fehlende Werte in OGD_gesausgaben01_HVD_HCHF_1_HEADER:


code           0
name           0
en_name        0
Unnamed: 3    16
Unnamed: 4    16
Unnamed: 5    16
Unnamed: 6    16
Unnamed: 7    16
Unnamed: 8    16
Unnamed: 9    16
dtype: int64


--- OGD_gesausgaben01_HVD_HCHF_1 ---

Erste 5 Zeilen:


C-ZEITGES-0   C-HCGES-0      F-HF1    F-HF11    F-HF12   F-HF121 F-HF122  \
0  ZEITGES-2004  HC-HC1-HC2  10808,144  3806,276  7001,868  7001,868   0,000   
1  ZEITGES-2004      HC-HC3   2440,161  2427,924    12,236    12,236   0,000   
2  ZEITGES-2004      HC-HC4    487,615    36,592   451,023   451,023   0,000   
3  ZEITGES-2004      HC-HC5   2604,742     3,820  2600,922  2600,922   0,000   
4  ZEITGES-2004      HC-HC6    378,498   293,059    85,439    85,439   0,000   

     F-HF2   F-HF21   F-HF22  F-HF23     F-HF3    F-HF31   F-HF32  F-HF4  \
0  963,953  821,200  142,753   0,000  2133,249  1965,609  167,640  0,000   
1   39,711    0,000   39,711   0,000   544,594   544,594    0,000  0,000   
2  133,839    0,000  133,839   0,000   110,471   105,254    5,218  0,000   
3   44,169   44,169    0,000   0,000  1673,625  1334,838  338,787  0,000   
4   47,713    0,000    7,916  39,797    79,907    79,907    0,000  0,000   

   F-ALLE_HF  
0  13905,347  
1   3024,465  
2    731,925  
3   4322,536  
4    506,117


Zufällige Stichprobe:


C-ZEITGES-0     C-HCGES-0      F-HF1     F-HF11     F-HF12    F-HF121  \
444  ZEITGES-2016    HC-HC1-HC2  17514,438   6437,355  11077,084  11077,084   
651  ZEITGES-2021        HC-HC2   1879,108    265,860   1613,248   1613,248   
112  ZEITGES-2007        HC-HC3   2810,512   2797,980     12,531     12,531   
532  ZEITGES-2018  HC-HC13-HC23   7496,009   1675,613   5820,395   5820,395   
692  ZEITGES-2022     HC-ALL_HC  38697,530  17456,301  21241,229  21241,229   

    F-HF122     F-HF2    F-HF21   F-HF22   F-HF23     F-HF3    F-HF31  \
444   0,000  1230,160  1034,891  195,269    0,000  3082,144  2770,602   
651   0,000   287,732    31,879  255,853    0,000   677,519   604,108   
112   0,000    58,167     0,000   58,167    0,000   655,381   655,381   
532   0,000   236,729   190,148   46,581    0,000  2630,688  2421,884   
692   0,000  3220,972  2271,108  829,258  120,607  7978,880  7206,073   

      F-HF32  F-HF4  F-ALLE_HF  
444  311,542  0,000  21826,742  
651   73,411  0,000   2844,359  
112    0,000  0,000   3524,060  
532  208,804  0,000  10363,426  
692  772,807  0,000  49897,383


Dateninfo:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 729 entries, 0 to 728
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   C-ZEITGES-0  729 non-null    category
 1   C-HCGES-0    729 non-null    category
 2   F-HF1        729 non-null    category
 3   F-HF11       729 non-null    category
 4   F-HF12       729 non-null    category
 5   F-HF121      729 non-null    category
 6   F-HF122      729 non-null    category
 7   F-HF2        729 non-null    category
 8   F-HF21       729 non-null    category
 9   F-HF22       729 non-null    category
 10  F-HF23       729 non-null    category
 11  F-HF3        729 non-null    category
 12  F-HF31       729 non-null    category
 13  F-HF32       729 non-null    category
 14  F-HF4        729 non-null    category
 15  F-ALLE_HF    729 non-null    category
dtypes: category(16)
memory usage: 217.2 KB

Deskriptive Statistiken (numerische und kategoriale Werte):


C-ZEITGES-0   C-HCGES-0  F-HF1 F-HF11 F-HF12 F-HF121 F-HF122  F-HF2  \
count            729         729    729    729    729     729     729    729   
unique            19          43    622    567    513     513       1    491   
top     ZEITGES-2021  HC-ALLE_HC  0,000  0,000  0,000   0,000   0,000  0,000   
freq              43          19     35     86    103     103     729    171   

       F-HF21 F-HF22 F-HF23  F-HF3 F-HF31 F-HF32  F-HF4 F-ALLE_HF  
count     729    729    729    729    729    729    729       729  
unique    380    248     20    534    515    217      1       672  
top     0,000  0,000  0,000  0,000  0,000  0,000  0,000    25,367  
freq      285    401    667    135    154    380    729         2


Anzahl eindeutiger Werte pro Spalte:
C-ZEITGES-0: 19 eindeutige Werte
C-HCGES-0: 43 eindeutige Werte
F-HF1: 622 eindeutige Werte
F-HF11: 567 eindeutige Werte
F-HF12: 513 eindeutige Werte
F-HF121: 513 eindeutige Werte
F-HF122: 1 eindeutige Werte
F-HF2: 491 eindeutige Werte
F-HF21: 380 eindeutige Werte
F-HF22: 248 eindeutige Werte
F-HF23: 20 eindeutige Werte
F-HF3: 534 eindeutige Werte
F-HF31: 515 eindeutige Werte
F-HF32: 217 eindeutige Werte
F-HF4: 1 eindeutige Werte
F-ALLE_HF: 672 eindeutige Werte
Fehlende Werte in OGD_gesausgaben01_HVD_HCHF_1:


C-ZEITGES-0    0
C-HCGES-0      0
F-HF1          0
F-HF11         0
F-HF12         0
F-HF121        0
F-HF122        0
F-HF2          0
F-HF21         0
F-HF22         0
F-HF23         0
F-HF3          0
F-HF31         0
F-HF32         0
F-HF4          0
F-ALLE_HF      0
dtype: int64


--- OGD_gesausgaben01_HVD_HCHF_1_C-HCGES-0 ---

Erste 5 Zeilen:


code                                               name   FK  \
0  HC-HC1-HC2  HC.1+HC.2 Leistungen der kurativen und rehabil...  NaN   
1      HC-HC3                                HC.3 Langzeitpflege  NaN   
2      HC-HC4     HC.4 Hilfsleistungen der Gesundheitsversorgung  NaN   
3      HC-HC5                            HC.5 Medizinische Güter  NaN   
4      HC-HC6                                    HC.6 Prävention  NaN   

                                             en_name  de_desc  de_link  \
0         HC.1+HC.2 Curative and rehabilitative care      NaN      NaN   
1                       HC.3 Long-term care (health)      NaN      NaN   
2  HC.4 Ancillary services (non-specified by func...      NaN      NaN   
3     HC.5 Medical goods (non-specified by function)      NaN      NaN   
4                               HC.6 Preventive care      NaN      NaN   

   en_desc  en_link  de_syn  en_syn Kategorieebene  
0      NaN      NaN     NaN     NaN            NaN  
1      NaN      NaN     NaN     NaN            NaN  
2      NaN      NaN     NaN     NaN            NaN  
3      NaN      NaN     NaN     NaN            NaN  
4      NaN      NaN     NaN     NaN            NaN


Zufällige Stichprobe:


code                                               name  \
28       HC-HC34                    HC.3.4 Häusliche Langzeitpflege   
9   HC-HC11-HC21  HC.1.1+HC.2.1 Stationäre kurative und rehabili...   
3         HC-HC5                            HC.5 Medizinische Güter   
15  HC-HC13-HC23  HC.1.3+HC.2.3 Ambulante kurative und rehabilit...   
4         HC-HC6                                    HC.6 Prävention   

            FK                                            en_name  de_desc  \
28      HC-HC3          HC.3.4 Home-based long-term care (health)      NaN   
9   HC-HC1-HC2  HC.1.1+HC.2.1 Inpatient curative and rehabilit...      NaN   
3          NaN     HC.5 Medical goods (non-specified by function)      NaN   
15  HC-HC1-HC2  HC.1.3+HC.2.3 Outpatient curative and rehabili...      NaN   
4          NaN                               HC.6 Preventive care      NaN   

    de_link  en_desc  en_link  de_syn  en_syn  Kategorieebene  
28      NaN      NaN      NaN     NaN     NaN  Unterkategorie  
9       NaN      NaN      NaN     NaN     NaN  Unterkategorie  
3       NaN      NaN      NaN     NaN     NaN             NaN  
15      NaN      NaN      NaN     NaN     NaN  Unterkategorie  
4       NaN      NaN      NaN     NaN     NaN             NaN


Dateninfo:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46 entries, 0 to 45
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   code            46 non-null     category
 1   name            46 non-null     category
 2   FK              39 non-null     category
 3   en_name         46 non-null     category
 4   de_desc         0 non-null      float64 
 5   de_link         0 non-null      float64 
 6   en_desc         0 non-null      float64 
 7   en_link         0 non-null      float64 
 8   de_syn          0 non-null      float64 
 9   en_syn          0 non-null      float64 
 10  Kategorieebene  39 non-null     object  
dtypes: category(4), float64(6), object(1)
memory usage: 7.4+ KB

Deskriptive Statistiken (numerische und kategoriale Werte):


de_desc  de_link  en_desc  en_link  de_syn  en_syn
count      0.0      0.0      0.0      0.0     0.0     0.0
mean       NaN      NaN      NaN      NaN     NaN     NaN
std        NaN      NaN      NaN      NaN     NaN     NaN
min        NaN      NaN      NaN      NaN     NaN     NaN
25%        NaN      NaN      NaN      NaN     NaN     NaN
50%        NaN      NaN      NaN      NaN     NaN     NaN
75%        NaN      NaN      NaN      NaN     NaN     NaN
max        NaN      NaN      NaN      NaN     NaN     NaN


Anzahl eindeutiger Werte pro Spalte:
code: 46 eindeutige Werte
name: 45 eindeutige Werte
FK: 7 eindeutige Werte
en_name: 46 eindeutige Werte
de_desc: 0 eindeutige Werte
de_link: 0 eindeutige Werte
en_desc: 0 eindeutige Werte
en_link: 0 eindeutige Werte
de_syn: 0 eindeutige Werte
en_syn: 0 eindeutige Werte
Kategorieebene: 1 eindeutige Werte
Fehlende Werte in OGD_gesausgaben01_HVD_HCHF_1_C-HCGES-0:


code               0
name               0
FK                 7
en_name            0
de_desc           46
de_link           46
en_desc           46
en_link           46
de_syn            46
en_syn            46
Kategorieebene     7
dtype: int64

### Ergebnisse:  

**1. Hauptdatensatz (Gesundheitsausgaben)**  
- **Anzahl:** 729 Zeilen, 16 Spalten.  
- **Datentypen:** Alle Spalten sind vom Typ `object`. Dies deutet darauf hin, dass möglicherweise numerische Werte als Text gespeichert wurden und eine Umwandlung notwendig sein könnte.  
- **Fehlende Werte:** Keine fehlenden Werte festgestellt.  
- **Mögliche Maßnahmen:** Überprüfung der Datenformate (z. B. Zahlen als Text) und mögliche Konvertierung in numerische Typen.  

---

**2. Gesundheitsleistungen und -güter (HC)**  
- **Anzahl:** 46 Zeilen, 10 Spalten.  
- **Fehlende Werte:**  
  - Spalten `de_desc`, `de_link`, `en_desc`, `en_link`, `de_syn`, `en_syn` enthalten ausschließlich fehlende Werte und sollten entfernt werden.  
  - Die Spalte `FK` weist 7 fehlende Werte auf, welche genauer überprüft werden müssen (z. B. Kategorie oder Obergruppe).  
- **Mögliche Maßnahmen:**  
  - Löschen irrelevanter Spalten mit ausschließlich fehlenden Werten.  
  - Umgang mit fehlenden Werten in `FK` (Ersatz, Gruppierung oder Löschung).  

---

**3. Zeitreihen der Gesundheitsausgaben**  
- **Anzahl:** 19 Zeilen, 10 Spalten.  
- **Fehlende Werte:**  
  - Die Spalte `Unnamed: 2` enthält nur `NaN`-Werte und sollte entfernt werden.  
  - Weitere Spalten (`de_desc`, `de_link`, etc.) sind ebenfalls vollständig leer.  
- **Datentypen:**  
  - Die Spalten `name` und `en_name` sind als `int64` klassifiziert – sollte überprüft werden, ob dies korrekt ist oder in `string`/`category` geändert werden muss.  
- **Mögliche Maßnahmen:**  
  - Entfernen leerer Spalten.  
  - Datentypen anpassen, falls erforderlich.  

---

**4. Header-Metadaten**  
- **Anzahl:** 16 Zeilen, 10 Spalten.  
- **Fehlende Werte:**  
  - Unbenannte Spalten `Unnamed: 3` bis `Unnamed: 9` enthalten ausschließlich fehlende Werte und können entfernt werden.  
- **Mögliche Maßnahmen:**  
  - Entfernen leerer Spalten zur Reduzierung unnötiger Daten.  

---

### Nächste Schritte:  
1. **Bereinigung der Daten:**  
   - Entfernen der vollständig leeren Spalten.  
   - Behandlung der fehlenden Werte in wichtigen Spalten (z. B. `FK`).  
   - Konvertierung der Spalten in geeignete Datentypen.  
   
2. **Speicherung des bereinigten Datensatzes** in einer Datei `korr.csv`.  

## 2. Datenbereinigung

**Beobachtungen zur Anzahl fehlender Werte:**  

- Der Hauptdatensatz enthält **keine** fehlenden Werte.  
- Der HC-Datensatz weist mehrere Spalten mit vollständig fehlenden Werten auf.  
- Unbenannte Spalten in den Zeitreihen- und Header-Daten enthalten nur NaN-Werte.  

**TODO:**  
- Entfernen unnötiger Spalten mit vollständig fehlenden Werten.  

In [165]:
df_hc_cleaned = df_hc.drop(columns=['de_desc', 'de_link', 'en_desc', 'en_link', 'de_syn', 'en_syn'])
df_time_cleaned = df_time.drop(columns=['Unnamed: 2', 'de_desc', 'de_link', 'en_desc', 'en_link', 'de_syn', 'en_syn'])
df_header_cleaned = df_header.drop(columns=['Unnamed: 3', 'Unnamed: 4', 'Unnamed: 5', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9'])

print("Bereinigter HC-Datensatz:")
display(df_hc_cleaned.sample(5))

print("Bereinigter Zeitreihen-Datensatz:")
display(df_time_cleaned.head())

print("Bereinigter Header-Datensatz:")
display(df_header_cleaned.head())

Bereinigter HC-Datensatz:


code                                               name          FK  \
30   HC-HC42               HC.4.2 Bildgebende Diagnoseverfahren      HC-HC4   
21   HC-HC23  HC.2.3 Ambulante rehabilitative Gesundheitsver...  HC-HC1-HC2   
28   HC-HC34                    HC.3.4 Häusliche Langzeitpflege      HC-HC3   
35  HC-HC513     HC.5.1.3 Sonstige medizinische Verbrauchsgüter      HC-HC5   
5     HC-HC7  HC.7 Verwaltung der Gesundheitsversorgung und ...         NaN   

                                              en_name  Kategorieebene  
30                            HC.4.2 Imaging services  Unterkategorie  
21              HC.2.3 Outpatient rehabilitative care  Unterkategorie  
28          HC.3.4 Home-based long-term care (health)  Unterkategorie  
35           HC.5.1.3 Other medical non-durable goods  Unterkategorie  
5   HC.7 Governance and health system and financin...             NaN

Bereinigter Zeitreihen-Datensatz:


code  name  en_name
0  ZEITGES-2004  2004     2004
1  ZEITGES-2005  2005     2005
2  ZEITGES-2006  2006     2006
3  ZEITGES-2007  2007     2007
4  ZEITGES-2008  2008     2008

Bereinigter Header-Datensatz:


code                                               name  \
0  C-ZEITGES-0                                               Zeit   
1    C-HCGES-0                   Gesundheitsleistungen_Güter (HC)   
2        F-HF1  HF.1 Finanzierungssystem Staat und verpflichte...   
3       F-HF11                   HF.1.1 Finanzierungssystem Staat   
4       F-HF12  HF.1.2 Verpflichtende Beitragssysteme: Kranken...   

                                             en_name  
0                                               Time  
1                         Health care functions (HC)  
2  HF.1 Government schemes and compulsory contrib...  
3                          HF.1.1 Government schemes  
4  HF.1.2 Compulsory contributory health insuranc...

### FK-Spalte (Fremdschlüssel)

Die FK-Spalte ist in einigen Zeilen leer (`NaN`), was darauf hindeutet, dass diese Zeilen übergeordnete Kategorien sind.  
Unser Ziel ist es, die Hierarchie der Gesundheitsleistungen zu verstehen und ggf. geeignete Anpassungen vorzunehmen.

**Zu untersuchende Aspekte:**

1. Anzahl der `NaN`- und nicht-`NaN`-Werte in der FK-Spalte.  
2. Identifikation der übergeordneten Kategorien anhand der FK-Spalte.  

In [166]:
# Anzahl der fehlenden Werte in der FK-Spalte
missing_fk = df_hc['FK'].isnull().sum()
total_rows = len(df_hc)

print(f"Anzahl der fehlenden FK-Werte (Überkategorien): {missing_fk}")
print(f"Anzahl der vorhandenen FK-Werte (Unterkategorien): {total_rows - missing_fk}")

# Anzeigen der Zeilen mit fehlender FK (Überkategorien)
df_hc_missing_fk = df_hc[df_hc['FK'].isnull()]
display(df_hc_missing_fk[['code', 'name']])

# Anzeigen der Zeilen mit vorhandener FK (Unterkategorien)
df_hc_with_fk = df_hc[df_hc['FK'].notnull()]
display(df_hc_with_fk[['code', 'name', 'FK']].head(10))

Anzahl der fehlenden FK-Werte (Überkategorien): 7
Anzahl der vorhandenen FK-Werte (Unterkategorien): 39


code                                               name
0  HC-HC1-HC2  HC.1+HC.2 Leistungen der kurativen und rehabil...
1      HC-HC3                                HC.3 Langzeitpflege
2      HC-HC4     HC.4 Hilfsleistungen der Gesundheitsversorgung
3      HC-HC5                            HC.5 Medizinische Güter
4      HC-HC6                                    HC.6 Prävention
5      HC-HC7  HC.7 Verwaltung der Gesundheitsversorgung und ...
6   HC-ALL_HC    Alle HC - Alle Gesundheitsleistungen und -güter

code                                               name  \
7         HC-HC1  HC.1 Leistungen der kurativen Gesundheitsverso...   
8         HC-HC2  HC.2 Leistungen der rehabilitativen Gesundheit...   
9   HC-HC11-HC21  HC.1.1+HC.2.1 Stationäre kurative und rehabili...   
10       HC-HC11   HC.1.1 Stationäre kurative Gesundheitsversorgung   
11       HC-HC21  HC.2.1 Stationäre rehabilitative Gesundheitsve...   
12  HC-HC12-HC22  HC.1.2+HC.2.2 Tagesfälle der kurativen und reh...   
13       HC-HC12  HC.1.2 Tagesfälle der kurativen Gesundheitsver...   
14       HC-HC22  HC.2.2 Tagesfälle der rehabilitativen Gesundhe...   
15  HC-HC13-HC23  HC.1.3+HC.2.3 Ambulante kurative und rehabilit...   
16       HC-HC13    HC.1.3 Ambulante kurative Gesundheitsversorgung   

            FK  
7   HC-HC1-HC2  
8   HC-HC1-HC2  
9   HC-HC1-HC2  
10  HC-HC1-HC2  
11  HC-HC1-HC2  
12  HC-HC1-HC2  
13  HC-HC1-HC2  
14  HC-HC1-HC2  
15  HC-HC1-HC2  
16  HC-HC1-HC2

**Ergebnisse der FK-Analyse:**  

- Einträge mit fehlender FK markieren übergeordnete Kategorien.  
- Einträge mit vorhandener FK verweisen auf ihre jeweilige Oberkategorie.  
- Diese Hierarchie hilft uns, die Daten später besser zu analysieren und zu gruppieren.  

**Nächster Schritt:**  
- Erstellung einer neuen Spalte `Kategorieebene`, die zwischen übergeordneten und untergeordneten Kategorien unterscheidet.  
- Falls sinnvoll, Speicherung einer hierarchischen Struktur für spätere Analysen oder Visualisierungen.  


In [167]:
# Neue Spalte zur Kategorisierung hinzufügen (Über-/Unterkategorie)
df_hc['Kategorieebene'] = df_hc['FK'].apply(lambda x: 'Oberkategorie' if pd.isnull(x) else 'Unterkategorie')

# anzeigen der kartegorieebene
display(df_hc[['code', 'name', 'FK', 'Kategorieebene']].head(15))

# anzahl der Ober- und Unterkategorien anzeigen
print(df_hc['Kategorieebene'].value_counts())


code                                               name  \
0     HC-HC1-HC2  HC.1+HC.2 Leistungen der kurativen und rehabil...   
1         HC-HC3                                HC.3 Langzeitpflege   
2         HC-HC4     HC.4 Hilfsleistungen der Gesundheitsversorgung   
3         HC-HC5                            HC.5 Medizinische Güter   
4         HC-HC6                                    HC.6 Prävention   
5         HC-HC7  HC.7 Verwaltung der Gesundheitsversorgung und ...   
6      HC-ALL_HC    Alle HC - Alle Gesundheitsleistungen und -güter   
7         HC-HC1  HC.1 Leistungen der kurativen Gesundheitsverso...   
8         HC-HC2  HC.2 Leistungen der rehabilitativen Gesundheit...   
9   HC-HC11-HC21  HC.1.1+HC.2.1 Stationäre kurative und rehabili...   
10       HC-HC11   HC.1.1 Stationäre kurative Gesundheitsversorgung   
11       HC-HC21  HC.2.1 Stationäre rehabilitative Gesundheitsve...   
12  HC-HC12-HC22  HC.1.2+HC.2.2 Tagesfälle der kurativen und reh...   
13       HC-HC12  HC.1.2 Tagesfälle der kurativen Gesundheitsver...   
14       HC-HC22  HC.2.2 Tagesfälle der rehabilitativen Gesundhe...   

            FK  Kategorieebene  
0          NaN             NaN  
1          NaN             NaN  
2          NaN             NaN  
3          NaN             NaN  
4          NaN             NaN  
5          NaN             NaN  
6          NaN             NaN  
7   HC-HC1-HC2  Unterkategorie  
8   HC-HC1-HC2  Unterkategorie  
9   HC-HC1-HC2  Unterkategorie  
10  HC-HC1-HC2  Unterkategorie  
11  HC-HC1-HC2  Unterkategorie  
12  HC-HC1-HC2  Unterkategorie  
13  HC-HC1-HC2  Unterkategorie  
14  HC-HC1-HC2  Unterkategorie

Kategorieebene
Unterkategorie    39
Name: count, dtype: int64


### Bereinigung der Sprach- und Null-Spalten

In den Datensätzen sind Spalten in deutscher und englischer Sprache enthalten.  
Da die englischen Spalten redundant sind, werden sie entfernt.  
Zusätzlich werden alle vollständig leeren Spalten identifiziert und aus den Daten gelöscht.

#### Schritte:
1. Entfernen der englischen Spalten.
2. Identifikation von komplett leeren Spalten.
3. Löschen der leeren Spalten, um die Datenqualität zu verbessern.
4. Überprüfung der Datenstruktur nach der Bereinigung.


In [168]:
def clean_df(dataframe, name):
    print(f"\n--- Bereinigung für {name} ---")

    # 1. Englische Spalten entfernen (beginnen mit 'en_')
    df_cleaned = df.drop(columns=[col for col in df.columns if col.startswith('en_')])

    # 2. Vollständig leerer Spalten
    empty_cols = df_cleaned.columns[df_cleaned.isnull().all()].tolist()
    print(f"Leere Spalten in {name}: {empty_cols}")

    # 3. Entfernen ovn leeren Spalten
    df_cleaned.drop(columns=empty_cols, inplace=True)

    # 4. Shape (Rows, Cols) nach Bereinigung
    print(f"Nach Bereinigung der englischen und leeren Spalten - {name}:")
    print(f"{name} shape:", df_cleaned.shape)

    # 5. Speichern der bereinigten DataFrames
    file_name = name + "_corr.csv"
    df_cleaned.to_csv(f'data/corr/{file_name}', index=False, sep=';')
    print(f"Bereinigte Datei für {name} gespeichert als {file_name}")

for name, df in dataframes.items():
    clean_df(df, name)


--- Bereinigung für OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0 ---
Leere Spalten in OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0: ['Unnamed: 2', 'de_desc', 'de_link', 'de_syn']
Nach Bereinigung der englischen und leeren Spalten - OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0:
OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0 shape: (19, 2)
Bereinigte Datei für OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0 gespeichert als OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0_corr.csv

--- Bereinigung für OGD_gesausgaben01_HVD_HCHF_1_HEADER ---
Leere Spalten in OGD_gesausgaben01_HVD_HCHF_1_HEADER: ['Unnamed: 3', 'Unnamed: 4', 'Unnamed: 5', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9']
Nach Bereinigung der englischen und leeren Spalten - OGD_gesausgaben01_HVD_HCHF_1_HEADER:
OGD_gesausgaben01_HVD_HCHF_1_HEADER shape: (16, 2)
Bereinigte Datei für OGD_gesausgaben01_HVD_HCHF_1_HEADER gespeichert als OGD_gesausgaben01_HVD_HCHF_1_HEADER_corr.csv

--- Bereinigung für OGD_gesausgaben01_HVD_HCHF_1 ---
Leere Spalten in OGD_g

### Mergen der DataFrames und Datenvorbereitung

Um die verschiedenen DataFrames mit relevanten Informationen zu kombinieren, haben wir alle DataFrames anhand des gemeinsamen Schlüssels `code` zusammengeführt. Dies stellt sicher, dass alle relevanten Daten in einem einzigen DataFrame vorliegen und keine Informationen verloren gehen.

#### Schritte:
1. **Mergen der DataFrames**:
   - Wir haben die DataFrames `df_main`, `df_hc`, `df_time` und `df_header` mithilfe des gemeinsamen Schlüssels `code` zusammengeführt.
   - Ein `left join` wurde verwendet, um sicherzustellen, dass alle Zeilen aus dem Hauptdatensatz (`df_main`) erhalten bleiben, und nur passende Zeilen aus den anderen DataFrames hinzugefügt werden.

2. **Datenvorbereitung (Preprocessing)**:
   - **Extraktion der numerischen Werte**: Alle numerischen Spalten (z.B. `int64`, `float64`) wurden extrahiert und in einer separaten CSV-Datei gespeichert.
   - **Normalisierung der numerischen Werte**: Die numerischen Werte wurden mit einer Min-Max-Normalisierung skaliert und ebenfalls in einer separaten CSV-Datei gespeichert.
   - **Extraktion der nominalen (kategorischen) Werte**: Alle nominalen Werte (z.B. `object`, `category`) wurden extrahiert und ebenfalls in einer separaten CSV-Datei gespeichert.

#### Ergebnis:
- **df_merged_num.csv**: Enthält die numerischen Werte aus allen zusammengeführten DataFrames.
- **df_merged_norm.csv**: Enthält die normalisierten numerischen Werte.
- **df_merged_nom.csv**: Enthält die nominalen (kategorischen) Werte.

Das Mergen und die Datenvorbereitung sorgen dafür, dass alle relevanten Daten an einem Ort vorhanden sind und für die spätere Analyse bereitgestellt werden.


In [178]:
# Daten einlesen
df_main = pd.read_csv('data/corr/OGD_gesausgaben01_HVD_HCHF_1_corr.csv', delimiter=';')
df_hc = pd.read_csv('data/corr/OGD_gesausgaben01_HVD_HCHF_1_C-HCGES-0_corr.csv', delimiter=';')
df_time = pd.read_csv('data/corr/OGD_gesausgaben01_HVD_HCHF_1_C-ZEITGES-0_corr.csv', delimiter=';')
df_header = pd.read_csv('data/corr/OGD_gesausgaben01_HVD_HCHF_1_HEADER_corr.csv', delimiter=';')

# 1. Mergen der DataFrames (angenommen, 'code' ist der gemeinsame Schlüssel)
merged_df = pd.merge(df_hc_corr, df_time_corr, on='code', how='left')  # 'inner' für den Schnitt der beiden DataFrames
final_merged_df = pd.merge(merged_df, df_header_corr, on='code', how='left')

# Überprüfen der ersten Zeilen des finalen DataFrames
display(final_merged_df)


# Funktion zur Bereinigung und Speicherung der Daten
def preprocess_and_save(df, filename_prefix):
    # 1. Numerische Werte extrahieren
    df_numeric = df.select_dtypes(include=['int64', 'float64'])

    # Speichern der numerischen Werte in 'num.csv'
    df_numeric.to_csv(f'{filename_prefix}_num.csv', index=False, sep=';')

    # 2. Normalisierung der numerischen Werte (Min-Max Normalisierung)
    df_normalized = (df_numeric - df_numeric.min()) / (df_numeric.max() - df_numeric.min())

    # Speichern der normalisierten Werte in 'norm.csv'
    df_normalized.to_csv(f'{filename_prefix}_norm.csv', index=False, sep=';')

    # 3. Nominale Werte extrahieren
    df_nominal = df.select_dtypes(include=['object', 'category'])

    # Speichern der nominalen Werte in 'nom.csv'
    df_nominal.to_csv(f'{filename_prefix}_nom.csv', index=False, sep=';')

# Anwendung auf das gemergte DataFrame
# preprocess_and_save(df_merged, 'df_merged')

print("Daten wurden erfolgreich gespeichert.")


name_x          FK  \
code                                                                          
HC-HC1-HC2    HC.1+HC.2 Leistungen der kurativen und rehabil...         NaN   
HC-HC3                                      HC.3 Langzeitpflege         NaN   
HC-HC4           HC.4 Hilfsleistungen der Gesundheitsversorgung         NaN   
HC-HC5                                  HC.5 Medizinische Güter         NaN   
HC-HC6                                          HC.6 Prävention         NaN   
HC-HC7        HC.7 Verwaltung der Gesundheitsversorgung und ...         NaN   
HC-ALL_HC       Alle HC - Alle Gesundheitsleistungen und -güter         NaN   
HC-HC1        HC.1 Leistungen der kurativen Gesundheitsverso...  HC-HC1-HC2   
HC-HC2        HC.2 Leistungen der rehabilitativen Gesundheit...  HC-HC1-HC2   
HC-HC11-HC21  HC.1.1+HC.2.1 Stationäre kurative und rehabili...  HC-HC1-HC2   
HC-HC11        HC.1.1 Stationäre kurative Gesundheitsversorgung  HC-HC1-HC2   
HC-HC21       HC.2.1 Stationäre rehabilitative Gesundheitsve...  HC-HC1-HC2   
HC-HC12-HC22  HC.1.2+HC.2.2 Tagesfälle der kurativen und reh...  HC-HC1-HC2   
HC-HC12       HC.1.2 Tagesfälle der kurativen Gesundheitsver...  HC-HC1-HC2   
HC-HC22       HC.2.2 Tagesfälle der rehabilitativen Gesundhe...  HC-HC1-HC2   
HC-HC13-HC23  HC.1.3+HC.2.3 Ambulante kurative und rehabilit...  HC-HC1-HC2   
HC-HC13         HC.1.3 Ambulante kurative Gesundheitsversorgung  HC-HC1-HC2   
HC-HC131      HC.1.3.1 Allgemeine ambulante Gesundheitsverso...  HC-HC1-HC2   
HC-HC132      HC.1.3.2 Ambulante zahnärztliche Gesundheitsve...  HC-HC1-HC2   
HC-HC133      HC.1.3.3 Spezialisierte ambulante Gesundheitsv...  HC-HC1-HC2   
HC-HC139      HC.1.3.9 Sonstige ambulante kurative Gesundhei...  HC-HC1-HC2   
HC-HC23       HC.2.3 Ambulante rehabilitative Gesundheitsver...  HC-HC1-HC2   
HC-HC14-HC24  HC.1.4+HC.2.4 Häusliche kurative und rehabilit...  HC-HC1-HC2   
HC-HC14         HC.1.4 Häusliche kurative Gesundheitsversorgung  HC-HC1-HC2   
HC-HC24       HC.2.4 Häusliche rehabilitative Gesundheitsver...  HC-HC1-HC2   
HC-HC31                        HC.3.1 Stationäre Langzeitpflege      HC-HC3   
HC-HC32                    HC.3.2 Tagesfälle der Langzeitpflege      HC-HC3   
HC-HC33                         HC.3.3 Ambulante Langzeitpflege      HC-HC3   
HC-HC34                         HC.3.4 Häusliche Langzeitpflege      HC-HC3   
HC-HC41                                 HC.4.1 Klinische Labore      HC-HC4   
HC-HC42                    HC.4.2 Bildgebende Diagnoseverfahren      HC-HC4   
HC-HC43                              HC.4.3 Patiententransporte      HC-HC4   
HC-HC51       HC.5.1 Arzneimittel und sonstige medizinische ...      HC-HC5   
HC-HC511                    HC.5.1.1 Verschriebene Arzneimittel      HC-HC5   
HC-HC512                      HC.5.1.2 Rezeptfreie Arzneimittel      HC-HC5   
HC-HC513         HC.5.1.3 Sonstige medizinische Verbrauchsgüter      HC-HC5   
HC-HC52       HC.5.2 Therapeutische Hilfsmittel und andere m...      HC-HC5   
HC-HC61                HC.6.1 Information, Bildung und Beratung      HC-HC6   
HC-HC62                          HC.6.2 Immunisierungsprogramme      HC-HC6   
HC-HC63                          HC.6.3 Früherkennungsprogramme      HC-HC6   
HC-HC64       HC.6.4 Programme zur Überwachung des Gesundhei...      HC-HC6   
HC-HC65       HC.6.5 Epidemiologische Überwachung, Risiko- u...      HC-HC6   
HC-HC66       HC.6.6 Katastrophenvorbereitung und Notfallpro...      HC-HC6   
HC-HC71             HC.7.1 Verwaltung der Gesundheitsversorgung      HC-HC7   
HC-HC72       HC.7.2 Verwaltung der Gesundheitssystemfinanzi...      HC-HC7   
HC-ALLE_HC      Alle HC - Alle Gesundheitsleistungen und -güter   HC-ALL_HC   

              Kategorieebene  name_y name  
code                                       
HC-HC1-HC2               NaN     NaN  NaN  
HC-HC3                   NaN     NaN  NaN  
HC-HC4                   NaN     NaN  NaN  
HC-HC5                   NaN     NaN  NaN  


Daten wurden erfolgreich gespeichert.
